### PR TITLE
feat: vm client changes for Azure Stack Hub support

### PR DESCRIPTION
**Pull request by JesusAlvarezTorres**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

/kind feature

**What this PR does / why we need it**:
VM client changes for Azure Stack Hub support.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
Part of a series of changes to add Azure Stack Hub support to all supported ARM resource clients.

**Release note**:
```
feat: vm client changes for Azure Stack Hub support
```

### CODE REVIEW

**Comment by andyzhangx** [comment]
/ok-to-test

@JesusAlvarezTorres would you sign the CLA first?
cc @songjiaxun

**Comment by JesusAlvarezTorres** [comment]
> /ok-to-test
> 
> @JesusAlvarezTorres would you sign the CLA first?
> cc @songjiaxun

Thanks @andyzhangx , just signed the CLA

File: pkg/azureclients/vmclient/azure_vmclient.go [hunk 1]
```diff
@@ -71,6 +77,7 @@ func New(config *azclients.ClientConfig) *Client {
 		rateLimiterReader: rateLimiterReader,
 		rateLimiterWriter: rateLimiterWriter,
 		subscriptionID:    config.SubscriptionID,
+		cloudName:         config.CloudName,
```
Comment (andyzhangx) [inline pkg/azureclients/vmclient/azure_vmclient.go:80]: why add a new field here, what about use `config.CloudName` directly, refer to

https://github.com/kubernetes-sigs/cloud-provider-azure/blob/083cf6de2feb34b98873082c60b6cd0eaefc925b/pkg/azureclients/diskclient/azure_diskclient.go#L62-L64

File: pkg/azureclients/vmclient/azure_vmclient.go [hunk 2]
```diff
@@ -42,6 +43,7 @@ var _ Interface = &Client{}
 type Client struct {
 	armClient      armclient.Interface
 	subscriptionID string
+	cloudName      string
```
Comment (feiskyer) [inline pkg/azureclients/vmclient/azure_vmclient.go:46]: this is not used?

File: pkg/azureclients/vmclient/azure_vmclient.go (hunk 1 above)
Comment (JesusAlvarezTorres) [reply pkg/azureclients/vmclient/azure_vmclient.go:80]: for consistency, there are 2 clients (for example the interface client) that use an additional API version (the compute api version). In those cases the switch to the Azure Stack API version happens using this field as the config is not available on the method that uses the compute API version. I kept the implementation the same across all clients.

File: pkg/azureclients/vmclient/azure_vmclient.go (hunk 2 above)
Comment (JesusAlvarezTorres) [reply pkg/azureclients/vmclient/azure_vmclient.go:46]: It is used on the new unit test and in other clients that use the compute API version (interface client for example). As mentioned on my reply to @andyzhangx above, I added cloudName to all clients for consistency.

File: pkg/azureclients/vmclient/azure_vmclient.go (hunk 1 above)
Comment (andyzhangx) [reply pkg/azureclients/vmclient/azure_vmclient.go:80]: all clients are initialized from azure cloud config, what's the benefit to add a redundant field(`cloudName`), only for unit test?

https://github.com/kubernetes-sigs/cloud-provider-azure/blob/74926ac455ea163d9f0a3a7b3f3dd7c4332dc785/pkg/provider/azure.go#L557

File: pkg/azureclients/vmclient/azure_vmclient.go (hunk 2 above)
Comment (feiskyer) [reply pkg/azureclients/vmclient/azure_vmclient.go:46]: thanks for clarification, then it makes sense

File: pkg/azureclients/vmclient/azure_vmclient.go (hunk 1 above)
Comment (JesusAlvarezTorres) [reply pkg/azureclients/vmclient/azure_vmclient.go:80]: For unit test yes, and in this case for consistency with other clients like the interface client that actually do need the new field. For example the change in the interface client (already merged) uses the new field to check for the cloud name in a method outside of the initialization.
https://github.com/kubernetes-sigs/cloud-provider-azure/blob/9aae7de0440b8a87e29224ca94e9ef8c47f93e21/pkg/azureclients/interfaceclient/azure_interfaceclient.go#L190

Comment (andyzhangx) [reply pkg/azureclients/vmclient/azure_vmclient.go:80]: thanks for the clarification, lgtm.

**Review by andyzhangx — APPROVED** [review]
/lgtm
/approve

**Pull request merged**